packages/react-kit/src/components/Popover/__tests__/Popover.model.spec.ts

```diff
@@ -0,0 +1,351 @@
+import {
+	PopoverPlacement,
+	PopoverAlign,
+	TVerticalMoveProps,
+	TVerticalPosition,
+	movePopoverVertically,
+	THorizontalMoveProps,
+	THorizontalPosition,
+	movePopoverHorizontally,
+} from '../Popover.model';
+import { PartialKeys } from '@devexperts/utils/dist/object';
+
+const createMoveVertically: (
+	defaultProps: PartialKeys<TVerticalMoveProps, 'placement'>,
+) => (
+	props: Partial<TVerticalMoveProps> & Pick<TVerticalMoveProps, 'placement'>,
+) => TVerticalPosition | undefined = defaultProps => props =>
+	movePopoverVertically({ ...defaultProps, checkBounds: true, ...props });
+
+const createMoveHorizontally: (
+	defaultProps: PartialKeys<THorizontalMoveProps, 'placement'>,
+) => (
+	props: Partial<THorizontalMoveProps> & Pick<THorizontalMoveProps, 'placement'>,
+) => THorizontalPosition | undefined = defaultProps => props =>
+	movePopoverHorizontally({ ...defaultProps, checkBounds: true, ...props });
+
+const placementTop = PopoverPlacement.Top;
+const placementBottom = PopoverPlacement.Bottom;
+const placementLeft = PopoverPlacement.Left;
+const placementRight = PopoverPlacement.Right;
+
+const alignTop = PopoverAlign.Top;
+const alignBottom = PopoverAlign.Bottom;
+const alignLeft = PopoverAlign.Left;
+const alignRight = PopoverAlign.Right;
+const alignMiddle = PopoverAlign.Middle;
+const alignCenter = PopoverAlign.Center;
+
+describe('Popover.model', () => {
+	describe('move popover vertically', () => {
+		it('should open upward', () => {
+			const anchorTop = 50;
+			const popoverHeight = 20;
+			const top = anchorTop - popoverHeight;
+			const result: TVerticalPosition = {
+				placement: placementTop,
+				align: alignCenter,
+				top,
+			};
+			expect(
+				movePopoverVertically({
+					placement: placementTop,
+					align: alignCenter,
+					anchorTop,
+					anchorBottom: 55,
+					popoverHeight,
+					parentHeight: 105,
+					checkBounds: true,
+				}),
+			).toEqual(result);
+		});
+
+		it('should open downward', () => {
+			const anchorBottom = 55;
+			const top = anchorBottom;
+			const result: TVerticalPosition = {
+				placement: placementBottom,
+				align: alignCenter,
+				top,
+			};
+			expect(
+				movePopoverVertically({
+					placement: placementBottom,
+					align: alignCenter,
+					anchorTop: 50,
+					anchorBottom,
+					popoverHeight: 20,
+					parentHeight: 105,
+					checkBounds: true,
+				}),
+			).toEqual(result);
+		});
+
+		it('should try open upward and will open downward', () => {
```

Comment:
Case name should describe it in a more complete way, e.g: "Should open downward if there is not enough space on top to open upwards"